pluginconfig/token.go

```diff
@@ -63,113 +68,189 @@ func (t TokenDataObserverConfig) WellFormed() error {
 
 // Validate checks that the observer's config is semantically correct - fields are set correctly
 // depending on the config's type
-func (t TokenDataObserverConfig) Validate() error {
+func (t *TokenDataObserverConfig) Validate() error {
+	if err := t.WellFormed(); err != nil {
+		return err
+	}
 	if t.IsUSDC() {
 		return t.USDCCCTPObserverConfig.Validate()
 	}
 	return errors.New("unknown token data observer type " + t.Type)
 }
 
-func (t TokenDataObserverConfig) IsUSDC() bool {
+func (t *TokenDataObserverConfig) IsUSDC() bool {
 	return t.Type == USDCCCTPHandlerType
 }
 
-const USDCCCTPHandlerType = "usdc-cctp"
+// MarshalJSON is a custom JSON marshaller for TokenDataObserverConfig.
+// It constructs raw map based on provided type. Custom marshaller is needed because default golang marshaller
+// doesn't marshal clashing fields of pointer embeddings even if only one pointer is present and rest are set to nil
+func (t *TokenDataObserverConfig) MarshalJSON() ([]byte, error) {
```

Comment:
Why do we need to MarshallJSON? Config is only read from JSON, never persisted, no?

Reply:
This is a public type. Hopefully the tooling uses it.